src/rdb/rdb_raft.c

```diff
@@ -1310,11 +1279,6 @@ rdb_raft_cb_log_pop(raft_server_t *raft, void *arg, raft_entry_t *entry,
 		db->d_lc_record.dlr_tail = tail;
 		return rc;
 	}
-	d_rank_list_free(db->d_replicas);
```

Comment:
no change necessarily requested here. It may only be a comprehension question.

just trying to follow the flow. In the new raft code, as we pop entries from the log, we will "undo" the membership change effects of the entries one-by-one. For example with an entry of type RAFT_LOGTYPE_ADD_NODE, we have to remove the node. So raft_pop_log() invokes raft_remove_node() that calls notify_membership_event callback aka rdb_raft_cb_... (taking care of freeing the rdb_raft_node in volatile memory), and then removes the raft_node_t from raft_server_private_t.nodes array in volatile memory.

But where does the rdb persistent list of replicas get updated in this flow? Maybe in a log offer callback rdb_raft_cb_log_offer() because the popping is being done as a step in the larger context of appending entries to the log e.g., raft_recv_appendentries()?

Reply:
Good question.

The persistent list of replicas ("list" below) is stored in a VOS container, the LC. There are two equivalent perspectives we can adopt. One is to view the LC as a log, with entries indexed by epochs. Hence when discarding all epochs >= i, we delete the "list" in entry i, but leave any "list" in entries < i intact. The other perspective is to view the LC as a multi-version storage, with Raft entries stored as versions. When discarding all epochs >= i, we therefore delete all "list" versions >= i, but leave the versions < i intact. If we read the "list" at this time, we get the last version of the "list" <= i - 1.

Yet another way to see it is that the "list" is similar to the KVSs in the sense that they can both be regarded as part of the "user state machine". So if popping doesn't need to do anything extra for the KVSs, it also doesn't need to do anything extra for the "list".

My intention was not to defer to offer time in the same AE. That would probably be too implicit a dependency.

Did I understand the question correctly?

Reply:
Yes I think so. Now if I comprehend better, the "undo" behavior of delete/pop involves changing the volatile state one-by-one (entry) while vos due to its versioned history of rdb_[n]replicas can get back to the persistent state for epochs < i with a single discard operation.

Reply:
Agreed.